File: Difficulty: Medium/Diagonal Tree Traversal/diagonal-tree-traversal.cpp
/* A binary tree node
struct Node
{
    int data;
    Node* left, * right;
}; */

class Solution {
  public:
    vector<int> diagonal(Node *root) {
        // code here
        vector<int> ans;
        if(!root) return ans;
        queue<Node*> que;
        que.push(root);
        while(!que.empty()) {
            Node* curNode = que.front();
            que.pop();
            while(curNode) {
                ans.push_back(curNode->data);
                if(curNode->left) que.push(curNode->left);
                curNode = curNode->right;
            }
            
        }
        return ans;
    }
};

File: Difficulty: Medium/Diagonal Tree Traversal/diagonal-tree-traversal.java
/* Node is defined as
class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data = data;
        left=null;
        right=null;
    }
}*/
class Tree {
    
    private void solve(Node root,int level, ArrayList<ArrayList<Integer>> ians) {
        if(root==null) return;
        
        if(ians.size()==level) {
            ians.add(new ArrayList<Integer> ());
        }
        ians.get(level).add(root.data);
        
        solve(root.left,level+1,ians);
        solve(root.right,level,ians);
        
        
    }
    
    public ArrayList<Integer> diagonal(Node root) {
        // add your code here.
        ArrayList<ArrayList<Integer>> ians = new ArrayList<>();
        ArrayList<Integer> ans = new ArrayList<>();
        int level = 0;
        
        solve(root,level,ians);
        
        for(ArrayList<Integer> jans:ians) {
            ans.addAll(jans);
        }
        
        return ans;
    }
}